[Accessibility] Make the rating stars accessible

Manussakis opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #51918 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to a product page.
2. Scroll down and open the reviews tab.
3. The rating stars behave now as a radio group, this means that:
    - If no star is selected, the focus ring should go to the first star.
    - Pressing tab if any star is focused should move the focus away from the stars group.
    - Once a star is focused, use the arrows on the keyboard to navigate through the stars and select them.
    - Screen readers should announce the focused star.
4. Verify the review functionality is working as before.
5. Verify the visual appearance of the rating stars didn't change per theme.
6. Test it across all Twenty* WP themes and Storefront.
7. **HEADS-UP**: The Storefront theme doesn't load Woo core stylesheet, so there is a visual regression that will need to be fixed on the Storefront code base.

**Before**:
![Screenshot 2024-11-22 at 17 22 48](https://github.com/user-attachments/assets/ec4a3f58-e858-4dc1-937b-f518eefcbe96)

**After**
![Screenshot 2024-11-22 at 17 23 46](https://github.com/user-attachments/assets/fd6a545c-af8b-4233-bb79-cec166476523)



<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [x] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Manussakis committed:
Add i18n_rating_legend and i18n_rating_options props to wc_single_product_params global JS object

Manussakis committed:
Add alternative text to star pseudo element

Manussakis committed:
Add keyboard navigation to rating starts

Manussakis committed:
Preserve old rating stars markup

Manussakis committed:
Create reset-rating-fieldset mixin

Manussakis committed:
Update rating style on TT3 theme

Manussakis committed:
Update rating style on TT2 theme

Manussakis committed:
Update rating style on TT1 theme

Manussakis committed:
Update rating style on TT theme

Manussakis committed:
Update rating style on T19 theme

Manussakis committed:
Update rating style on T17 theme

Manussakis committed:
Adjust unit to rem

Manussakis commented on plugins/woocommerce/client/legacy/css/_mixins.scss:
```diff
@@ -383,3 +383,15 @@
 }
+
+@mixin reset-rating-fieldset {
+	fieldset {
+		border: 0;
+		margin: 0;
+		padding: 0;
+
+		legend {
+			padding: 0;
+		}
+	}
```
Remove `fieldset` default styles.
![image](https://github.com/user-attachments/assets/e72fab19-dc0f-477a-9ca0-d424cd8ef7d9)

Manussakis commented on plugins/woocommerce/client/legacy/css/twenty-nineteen.scss:
```diff
@@ -506,7 +506,23 @@ table.variations {
+		.comment-form-rating {
+			@include reset-rating-fieldset();
+
+			fieldset {
```
Once the `fieldset` HTML element was added, each Twenty* theme required adjustments to preserve the old visual appearance.

Manussakis commented on plugins/woocommerce/client/legacy/css/woocommerce.scss:
```diff
@@ -858,6 +862,7 @@ p.demo_store,
 				line-height: 1;
 				font-family: "WooCommerce";
 				content: "\e021";
+				content: "\e021" / "";
```
`/ ""` is an alternative text for screen readers so they don't try to announce the "\e021" content.

**Reference**:
- https://www.w3.org/TR/css-content-3/#alt

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -46,18 +46,32 @@ jQuery( function( $ ) {
-							<a class="star-5" href="#">5</a>\
-						</span>\
-					</p>'
+					'<fieldset>\
```
The rating stars behave now like a radio group. To mitigate regressions on other themes and plugins, I preserved the old markup:
```html
<p>
  <span>
    <a>...</a>
    <a>...</a>
    ...
  </span>
</p>
```

and only wrapped it with `fieldset` and added the `legend`.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -66,8 +80,14 @@ jQuery( function( $ ) {
 			$rating.val( $star.text() );
-			$star.siblings( 'a' ).removeClass( 'active' );
-			$star.addClass( 'active' );
+			$star.siblings( 'a' )
+				.removeClass( 'active' )
+				.attr( 'aria-checked', 'false' )
+				.attr( 'tabindex', '-1' );
+			$star
+				.addClass( 'active' )
+				.attr( 'aria-checked', 'true' )
+				.attr( 'tabindex', '0' );
```
When the rating group is focused, the focus ring appears only on the selected option, just like in a normal radio group.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -81,6 +101,26 @@ jQuery( function( $ ) {
 				return false;
 			}
+		} )
+		.on( 'keydown', '#respond p.stars a', function( e ) {
```
Add keyboard navigation with the arrow keys.

Manussakis commented on plugins/woocommerce/includes/class-wc-frontend-scripts.php:
```diff
@@ -508,6 +508,14 @@ private static function get_script_data( $handle ) {
 				$params = array(
 					'i18n_required_rating_text'         => esc_attr__( 'Please select a rating', 'woocommerce' ),
+					'i18n_rating_legend'                => esc_attr__( 'Your rating', 'woocommerce' ),
+					'i18n_rating_options'               => array(
+						esc_attr__( '1 of 5 stars', 'woocommerce' ),
+						esc_attr__( '2 of 5 stars', 'woocommerce' ),
+						esc_attr__( '3 of 5 stars', 'woocommerce' ),
+						esc_attr__( '4 of 5 stars', 'woocommerce' ),
+						esc_attr__( '5 of 5 stars', 'woocommerce' ),
+					),
```
Improve how stars are announced by screen readers.

Manussakis committed:
Add changelog file

Manussakis committed:
Merge branch 'trunk' into fix/51918

Manussakis committed:
Fix start selection

Manussakis committed:
Merge branch 'trunk' into fix/51918

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -46,28 +46,49 @@ jQuery( function( $ ) {
-			$rating.val( $star.text() );
-			$star.siblings( 'a' ).removeClass( 'active' );
-			$star.addClass( 'active' );
+			$rating.val( starPos );
```
Set the rating value by getting the star position in the DOM instead of relying on the text value.

dkotter reviewed (approved):
@ankitguptaindia @qasumitbagthariya This is ready for QA

qasumitbagthariya commented:
QA update: ⚠️
---

@Manussakis 

When using the keyboard's left and right arrows to navigate through the stars, the product image also scrolls. This behavior should not occur, as the arrow keys should exclusively focus on star navigation without affecting the product image.


https://github.com/user-attachments/assets/9fc48ebb-c2ef-49a0-856f-57032142597d

Manussakis committed:
Merge branch 'trunk' into fix/51918

Manussakis committed:
Prevent keyup events from being triggered on the flexslider when the focus is on the stars or tabs.

Manussakis commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -81,11 +102,46 @@ jQuery( function( $ ) {
 				return false;
 			}
+		} )
+		.on( 'keyup', function() {
+			var $activeElement = $( document.activeElement );
+			var sliderKeyupBlockers = [ '.stars', '.tabs', '.wc-tabs'];
+			var $closestBlocker = $activeElement.closest( sliderKeyupBlockers.join( ', ' ) );
+
+			if ( $closestBlocker.length ) {
+				// Prevent keyup events from being triggered on the flexslider when the focus is on the stars or tabs.
+				productGalleryElement.data('flexslider').animating = true;
+				return;
+			}
+
+			productGalleryElement.data('flexslider').animating = false;
+		})
```
When the focus is on an element inside the `.stars` or `.tabs` wrappers, we want the arrow keys to move across their nested elements and not to move the product slider.
Looking into the [Flexslider plugin](https://github.com/woocommerce/FlexSlider/blob/master/jquery.flexslider.js#L109), I realized that setting `animating` to `true` would do the trick.

Manussakis commented:
Good catch @qasumitbagthariya , thanks!

@dkotter Please review this commit that fixes the issue flagged by @qasumitbagthariya .
- [Prevent keyup events from being triggered on the flexslider when the …](https://github.com/woocommerce/woocommerce/pull/53018/commits/7967f0e2cb70682079944abfe832eeb2dfda777d)

Thanks!

dkotter commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -81,11 +102,46 @@ jQuery( function( $ ) {
+		} )
+		.on( 'keyup', function() {
+			var $activeElement = $( document.activeElement );
+			var sliderKeyupBlockers = [ '.stars', '.tabs', '.wc-tabs'];
```
```suggestion
			var sliderKeyupBlockers = [ '.stars', '.tabs', '.wc-tabs' ];
```

dkotter commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -81,11 +102,46 @@ jQuery( function( $ ) {
+
+			if ( $closestBlocker.length ) {
+				// Prevent keyup events from being triggered on the flexslider when the focus is on the stars or tabs.
+				productGalleryElement.data('flexslider').animating = true;
```
```suggestion
				productGalleryElement.data( 'flexslider' ).animating = true;
```

dkotter commented on plugins/woocommerce/client/legacy/js/frontend/single-product.js:
```diff
@@ -81,11 +102,46 @@ jQuery( function( $ ) {
+				return;
+			}
+
+			productGalleryElement.data('flexslider').animating = false;
```
```suggestion
			productGalleryElement.data( 'flexslider' ).animating = false;
```

qasumitbagthariya commented:
QA update: ⚠️
---

@Manussakis Thank you for the quick fix! 

The scrolling issue is resolved, but a new issue has emerged. When using the keyboard's left and right arrows to navigate through the stars, and clicking on the product image afterward the images can't changing. Please check the issue in the attached video for reference.




https://github.com/user-attachments/assets/7b0676c8-a6a4-4300-a3e5-90a143409169

Manussakis committed:
Add white space to keep the code standard

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

Manussakis committed:
Add white space to keep the code standard

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

Manussakis committed:
Add white space to keep the code standard

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

Manussakis committed:
Fix slider when clicking on the thumbnail with focus on tab or stars

Manussakis commented:
@dkotter ,
@qasumitbagthariya is on fire and this PR needs another round of code review.
- [Fix slider when clicking on the thumbnail with focus on tab or stars](https://github.com/woocommerce/woocommerce/pull/53018/commits/38c6f37239cf4b9db7345dc625633725c315a084)

Thanks!

qasumitbagthariya commented:
### QA/Test Report ✅ 
-------------

I have checked this PR in the `Manussakis:fix/51918` branch and the changes are working as expected.

https://github.com/user-attachments/assets/c3b605f4-c946-4992-89a0-bcc1dba32828


**Testing Environment** 

<Details>

- WordPress: 6.7
- Theme: Storefront 4.6.0
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS Ventura 13.3 
- Git Branch: Manussakis:fix/51918
</Details>

@Manussakis  please remove [WIP] from the title and take this PR out of draft.

cc @vikrampm1

vikrampm1 commented:
@louwie17 this PR is ready for your review!

vikrampm1 commented:
@louwie17 checking back on the request for review here.

Manussakis committed:
Merge branch 'trunk' into fix/51918

louwie17 commented:
> @louwie17 checking back on the request for review here.

Hey @vikrampm1 sorry for missing the first ping. I plan on reviewing this on Monday.

louwie17 reviewed (commented):
Hey @Manussakis, thanks for working on this, the changes overall look good to me and it tested well.
The only caveat is that it does change the structure of the rating field and would break themes that don't use the `woocommerce.scss` file (https://github.com/woocommerce/woocommerce/blob/trunk/docs/theme-development/classic-theme-developer-handbook.md#disabling-woocommerce-styles  like `Astra` for example ). I wonder if there is a better way to have themes opt-in to this? Like using the [`add_theme_support`](https://developer.wordpress.org/reference/functions/add_theme_support/) function for example, but not sure if that is overkill.

Just curious what others in @woocommerce/woo-fse think?

louwie17 commented:
@Manussakis to follow up on this, could we also address the accessibility without changing the DOM structure? By keeping the `keydown` listener and handling the screen reader in another way? using `title` on the link? 
Given this also causes the regression in StoreFront, which is still heavily used, I did rather keep any DOM changes minimal.

Having said that, I do think this would be a better long term solution, and we could add this as the new default by the use of `add_theme_support` or something similar. So themes can opt-in and update when they feel ready.

Manussakis commented:
@louwie17 Thanks for the feedback and directions!
I will update the PR and request your review once it's done.

Thanks!

Manussakis committed:
Merge branch 'trunk' into fix/51918

Manussakis committed:
Remove styles for fieldset inside form rating

Manussakis committed:
Remove i18n_rating_legend translation

Manussakis committed:
Add id to the rating form label

Manussakis committed:
Remove fieldset and legend from rating group

Manussakis commented:
@louwie17 I figured out another way to achieve the same result without changing the DOM structure. Therefore, without causing regressions, so implementing the `add_theme_support` functionality may not be necessary.

These are the main changes I made:

1. Removed the `<fieldset>` tag and added the `role="group"` attribute to the `<span>` element that wraps the links.
2. Added `id="comment-form-rating-label"` attribute to the rating form label.
3. Added the `aria-labelledby="comment-form-rating-label"` to the `<span>` element.

@dkotter This PR is ready for another round of code review.

Thanks!

Manussakis commented:
@qasumitbagthariya 
After implementing the changes highlighted in [this comment](https://github.com/woocommerce/woocommerce/pull/53018#issuecomment-2551360472), this PR now needs another round of internal QA.

Thanks!

louwie17 reviewed (approved):
> @louwie17 I figured out another way to achieve the same result without changing the DOM structure. Therefore, without causing regressions, so implementing the add_theme_support functionality may not be necessary.

Awesome, thanks @Manussakis!

It tested well and the code looks good, thanks for updating this in a manner that keeps the same HTML structure, nice work!!

Btw quick side note, while testing this I did notice this error, which happens on products that don't use the gallery. I believe this is related to changes in this PR: https://github.com/woocommerce/woocommerce/pull/53158 are you able by chance to fix this?
<img width="1508" alt="Screenshot 2024-12-19 at 10 29 07 AM" src="https://github.com/user-attachments/assets/3c051960-753b-43c3-8355-775ae2b26eb7" />

louwie17 commented:
> Btw quick side note, while testing this I did notice this error, which happens on products that don't use the gallery. I believe this is related to changes in this PR: #53158 are you able by chance to fix this? <img alt="Screenshot 2024-12-19 at 10 29 07 AM" width="1508" src="https://private-user-images.githubusercontent.com/2240960/397323828-3c051960-753b-43c3-8355-775ae2b26eb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ2MDQyMzQsIm5iZiI6MTczNDYwMzkzNCwicGF0aCI6Ii8yMjQwOTYwLzM5NzMyMzgyOC0zYzA1MTk2MC03NTNiLTQzYzMtODM1NS03NzVhZTJiMjZlYjcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMTlUMTAyNTM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzkwODU0YzgwOWU3ZTc1NTBiNDJjMmIxNTIwZTZhMjY5MTdmZTVjNTFkNTAwMGUzMDQxOGRmNGE0Yzg2N2RiNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vgfxqztQCGN9jiZm9W1s5hbV-MGJKkZPs1CNR7FUaJ0">

I end up pushing a fix up for this here: https://github.com/woocommerce/woocommerce/pull/53877 as it needs a code freeze exception for the 9.6 release.

Manussakis commented:
> I end up pushing a fix up for this here: https://github.com/woocommerce/woocommerce/pull/53877 as it needs a code freeze exception for the 9.6 release.

Thank you for fixing that, @louwie17 !

qasumitbagthariya commented:
QA/Test Report ✅
---

I have checked this PR in the Manussakis:fix/51918 branch and the changes are working as expected.


<img width="853" alt="image" src="https://github.com/user-attachments/assets/21e3e7a2-9240-4ad0-9af5-76fe4ae84574" />

https://github.com/user-attachments/assets/855575b9-8046-4f52-b46d-30c165aa7022


**Testing Environment** 

<Details>

- WordPress: 6.7.1
- Theme: TT5 1,0
- WooCommerce - WooCommerce - Dev version with this fix specific branch.
- PHP: 8.0.30
- Web Server: Nginx 1.20.2
- Browser: Chrome 
- OS: macOS 15.2
- Git Branch: Manussakis:fix/51918
</Details>


cc @vikrampm1

Manussakis committed:
Merge branch 'trunk' into fix/51918

Manussakis committed:
Update single-product-review template version

jeffpaul commented:
@louwie17 are there any other concerns here or is this ready to merge?

louwie17 commented:
> @louwie17 are there any other concerns here or is this ready to merge?

No, thanks for the re-ping and @Manussakis for updating with the latest `trunk`. I will merge this.
